File: java/backend/src/main/java/com/etour/app/controller/DepartureController.java
package com.etour.app.controller;

import java.time.LocalDate;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.*;

import com.etour.app.entity.DepartureDateMaster;
import com.etour.app.service.DepartureService;          
import com.etour.app.dto.DepartureDateDTO;
import org.springframework.http.ResponseEntity;


@RestController
@RequestMapping("/api/departures")
public class DepartureController {

    @Autowired
    private DepartureService service;

    // ✅ 1) Add Departure Date
    @PostMapping("/add")
    public DepartureDateMaster addDepartureDate(@RequestBody DepartureDateMaster departure) {
        return service.addDepartureDate(departure);
    }

    // ✅ 2) Update Departure Date
    @PutMapping("/update")
    public DepartureDateMaster updateDeparture(@RequestBody DepartureDateMaster departure) {
        return service.updateDeparture(departure);
    }

    // ✅ 3) Get All Departure Dates
    @GetMapping
    public List<DepartureDateMaster> getAllDepartures() {
        return service.getAllDepartures();
    }

    // ✅ 4) Get Departure By ID
    @GetMapping("/{id}")
    public DepartureDateMaster getDepartureById(@PathVariable int id) {
        return service.getDepartureById(id);
    }

    // ✅ 5) Delete Departure By ID
    @DeleteMapping("/delete/{id}")
    public ResponseEntity<String> deleteDepartureDate(@PathVariable Integer id) {
        service.deleteDepartureDate(id);
        return ResponseEntity.ok("Departure date deleted successfully!");
    }

    // ✅ 6) Get Departures By Category (catmaster_id)
    @GetMapping("/category/{catmasterId}")
    public ResponseEntity<List<DepartureDateDTO>> getDepartureDatesByCatmasterId(@PathVariable Integer catmasterId) {
        return ResponseEntity.ok(service.getDepartureDatesByCatmasterId(catmasterId));
    }

    // ✅ 7) Get Upcoming Departures (today onwards)
    @GetMapping("/upcoming")
    public List<DepartureDateMaster> getUpcomingDepartures() {
        return service.getUpcomingDepartures();
    }

    // ✅ 8) Get Departures Between Dates
    // Example:
    // /departure/between?start=2026-01-01&end=2026-12-31
    @GetMapping("/between")
    public List<DepartureDateMaster> getDeparturesBetween(
            @RequestParam("start") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate start,
            @RequestParam("end") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate end) {

        return service.getDeparturesBetween(start, end);
    }
}

File: dotnet/backend/Controllers/DepartureController.cs
using Microsoft.AspNetCore.Mvc;
using Etour_Backend_dotnet.Models;
using Etour_Backend_dotnet.Services;

namespace Etour_Backend_dotnet.Controllers;

[ApiController]
[Route("api/departures")]
public class DepartureController : ControllerBase
{
    private readonly IDepartureService _departureService;

    public DepartureController(IDepartureService departureService)
    {
        _departureService = departureService;
    }

    // ==========================
    // 1. ADD DEPARTURE DATE
    // ==========================
    [HttpPost("add")]
    public async Task<IActionResult> AddDepartureDate([FromBody] departure_date_master departure)
    {
        var result = await _departureService.AddDepartureDate(departure);
        return Ok(result);
    }

    // ==========================
    // 2. UPDATE DEPARTURE
    // ==========================
    [HttpPut("update")]
    public async Task<IActionResult> UpdateDeparture([FromBody] departure_date_master departure)
    {
        try
        {
            var result = await _departureService.UpdateDeparture(departure);
            return Ok(result);
        }
        catch (Exception ex)
        {
            return NotFound(new { message = ex.Message });
        }
    }

    // ==========================
    // 3. GET ALL DEPARTURES
    // ==========================
    [HttpGet]
    public async Task<IActionResult> GetAllDepartures()
    {
        var departures = await _departureService.GetAllDepartures();
        return Ok(departures);
    }

    // ==========================
    // 4. GET DEPARTURE BY ID
    // ==========================
    [HttpGet("{id}")]
    public async Task<IActionResult> GetDepartureById(int id)
    {
        try
        {
            var departure = await _departureService.GetDepartureById(id);
            return Ok(departure);
        }
        catch (Exception ex)
        {
            return NotFound(new { message = ex.Message });
        }
    }

    // ==========================
    // 5. DELETE DEPARTURE
    // ==========================
    [HttpDelete("delete/{id}")]
    public async Task<IActionResult> DeleteDepartureDate(int id)
    {
        try
        {
            await _departureService.DeleteDepartureDate(id);
            return Ok(new { message = "Departure date deleted successfully!" });
        }
        catch (Exception ex)
        {
            return NotFound(new { message = ex.Message });
        }
    }

    // ==========================
    // 6. GET BY CATEGORY
    // ==========================
    [HttpGet("category/{catmasterId}")]
    public async Task<IActionResult> GetDepartureDatesByCatmasterId(int catmasterId)
    {
        var departures = await _departureService.GetDepartureDatesByCatmasterId(catmasterId);
        return Ok(departures);
    }

    // ==========================
    // 7. GET UPCOMING DEPARTURES
    // ==========================
    [HttpGet("upcoming")]
    public async Task<IActionResult> GetUpcomingDepartures()
    {
        var departures = await _departureService.GetUpcomingDepartures();
        return Ok(departures);
    }

    // ==========================
    // 8. GET BETWEEN DATES
    // ==========================
    [HttpGet("between")]
    public async Task<IActionResult> GetDeparturesBetween(
        [FromQuery] DateOnly start,
        [FromQuery] DateOnly end)
    {
        var departures = await _departureService.GetDeparturesBetween(start, end);
        return Ok(departures);
    }
}
